### ISSUE TITLE
Allow self.log from on_validation_end

### ISSUE THREAD

**Issue by indigoviolet** [PyTorchLightning/pytorch-lightning#5285]
## 🚀 Feature

Currently, `self.log` is [not supported](https://github.com/PyTorchLightning/pytorch-lightning/blob/9c8701f2e2c09a984ac0322942c8f742a7cc6bc9/pytorch_lightning/trainer/connectors/logger_connector/callback_hook_validator.py#L194) in `on_validation_end`. I would like it to be.

### Motivation

I want to compute some metric each time the validation loop runs, but I don't want it to be tied to `on_validation_batch_end`: i.e I want to do my own batching. I also want to `self.log` this so I can then use it as a key to monitor for checkpointing. Finally, the documentation claims:

```
Use the log() method to log from anywhere in a LightningModule and Callback except functions with batch_start in their names.
```

I don't see why we should prevent logging in this method, it just adds a surprising friction.

### Pitch

Don't prevent logging in `on_validation_end`.

**Comment by awaelchli** [comment]
Hi, `on_validation_end` is the old name. Use `on_validation_epoch_end`.

**Comment by Svito-zar** [comment]
I would also like to see this feature

**Comment by tchaton** [comment]
Hey there,

It looks like a bug. I will look into it.
Sorry for the inconvenience.

Best,
T.C

**Comment by awaelchli** [comment]
the on_validation_end hook runs after the others (like on_validation_epoch_end) so what would be the expected behaviour for arguments set to `on_epoch=True`? It would be for the next epoch, right?

**Comment by tchaton** [comment]
Hey @awaelchli, expected arguments would be on_step=False, on_epoch=[False, True]

**Comment by collinmccarthy** [comment]
Is logging from `on_validation_end` supposed to be supported? When I try to log I'm getting:

```
pytorch_lightning.utilities.exceptions.MisconfigurationException: on_validation_end function doesn't support logging using self.log() yet.
```

I would like to log the wait count to the progress bar so I can see how close I am to early stopping. I'm calling this from my `pl.LightningModule` as follows, after storing the `EarlyStopping` callback object in `self._early_stop_callback`.

```
self.log(
    name='wait',
    value=self._early_stop_callback.wait_count,
    on_step=False,
    on_epoch=True,
    prog_bar=True,
    logger=False
)
```

If this isn't supposed to be supported, nor planned to be, is there a good work-around I'm not seeing? Right now I'm setting a flag in `on_validation_end` to check and log this the next time `on_train_batch_start` runs, but that is not ideal.

Thanks,
Collin

**Comment by collinmccarthy** [comment]
For logging to the progress bar I forgot I could update it directly via `get_progress_bar_dict()` but I still thinking logging `wait_count` from `on_validation_end` is most appropriate so having that supported would be ideal. Maybe this is the case with #4969?

**Comment by aleSuglia** [comment]
@tchaton I tried installing the latest version from pip and logging in `on_validation_end()` doesn't work. Ideas?

**Comment by carmocca** [comment]
Re-opening and assigning myself. We can add support for this now.

**Comment by mariomeissner** [comment]
What happened to this feature? @carmocca 

I'm implementing a callback that runs a special evaluation loop on a separate dataset. I want to log the result of that evaluation. It runs at the `on_evaluation_end` event, of course, since I want to run this after the normal evaluation loop.

Ideally I'd like to be able to just do `pl_module.log("special/val", accuracy.compute())`, but it raises:
```
pytorch_lightning.utilities.exceptions.MisconfigurationException: You can't `self.log()` inside `on_validation_end`
```

Until this feature is added, what can I do to get around this issue?

**Comment by carmocca** [comment]
> Until this feature is added, what can I do to get around this issue?

Can't you use `on_validation_epoch_end`?

**Comment by mariomeissner** [comment]
> Can't you use `on_validation_epoch_end`?
I could indeed. Thank you for the hint.

**Comment by carmocca** [comment]
Unfortunately this cannot be supported without caveats because values logged in `LightningModule.on_validation_end` will not be available for checkpointing since `ModelCheckpoint` is a callback and currently `Callback`s run before the `LightningModule`

I recommend that `on_validation_epoch_end` is used instead, and if you need to log an aggregation over multiple validation epochs within a training epoch (`val_check_interval<1.0`) then you will need to only log on the last `on_validation_epoch_end` call for the epoch to replicate logging in `on_validation_end`.

The only way to properly support this would be to change the hook order, but that would be a large breaking change.

More context: https://github.com/PyTorchLightning/pytorch-lightning/pull/11543#discussion_r789015322

**Comment by TrentBrick** [comment]
This sounds very overly complicated and is causing serious problems with my logging. Can this issue at least me displayed more prominently in the docs? Thanks.

**Comment by carmocca** [comment]
What exactly would you like to mention? PRs to improve docs are welcome if you have specific ideas.

Also, can you elaborate on what exactly are those problems? Logging is not allowed so it should be raising an exception.

**Comment by rohitgr7** [comment]
also hooks that support logging are mentioned here: https://pytorch-lightning.readthedocs.io/en/latest/extensions/logging.html#id3

**Comment by huidiwiwi** [comment]
Hi, I'm new to pytorch lightning, could you please give a example of how to achieve this? 

> you will need to only log on the last `on_validation_epoch_end` call for the epoch to replicate logging in `on_validation_end`.

(I try to track the best val metrics with MetricsTracker, by logging the tracker.best_metric() on hook `on_validation_epoch_end`, I get a best metric on current state. How could I keep only the final one in logs? )
thx a lot.

**Comment by huidiwiwi** [comment]
> > How could I keep only the final one in logs? )
> 
> what do you mean by keeping the final one in logs? didn't get you here.

Hi rohitgr7, while I am tracking my validation metrics, I would like to use self.log() to log the tracker.best_metric() at the end of all train&validation loops.

**Comment by rohitgr7** [comment]
then just do?
```py
def on_validation_epoch_end(self):
    self.log('val_metric', tracker.best_metric())
```

**Comment by huidiwiwi** [comment]
> Hi, but this will log the best_metric of all epochs in the history instead only the last one. 
I want to get a single value instead of a list of values.

**Comment by rohitgr7** [comment]
okay.. you want to log it only once?

then just add a condition so that the log call is made only when the training is completed.
```py
def on_validation_epoch_end(self):
    if self.current_epoch == self.trainer.max_epochs - 1:
        self.log('val_metric', tracker.best_metric)
```
